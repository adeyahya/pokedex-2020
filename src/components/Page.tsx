import { FunctionComponent, ReactNode, memo } from 'react';
import { useRouter } from 'next/dist/client/router';
import Link from 'next/link';

interface Props {
  children: ReactNode;
}

export const Page: FunctionComponent<Props> = memo(({ children }) => {
  const router = useRouter();
  const isHomepage = router.pathname === '/';

  return (
    <div>
      <nav className="flex items-center justify-between flex-wrap bg-indigo-900 p-5">
        <div className="flex items-center flex-shrink-0 text-white mr-6">
          {!isHomepage && (
            <button onClick={router.back} className="lg:hidden mr-4">
              <svg
                viewBox="64 64 896 896"
                focusable="false"
                data-icon="arrow-left"
                width="1.4rem"
                height="1.4rem"
                fill="currentColor"
                aria-hidden="true"
              >
                <path d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"></path>
              </svg>
            </button>
          )}

          <Link href="/">
            <a>
              <span className="font-semibold text-xl tracking-tight">Pokedex</span>
            </a>
          </Link>
        </div>
      </nav>
      {children}
    </div>
  );
});
