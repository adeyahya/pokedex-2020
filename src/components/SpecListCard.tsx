import { FunctionComponent, memo } from 'react';

interface Props {
  list: React.ReactNode[];
  label: string;
  hideLabel?: boolean;
}
export const SpecListCard: FunctionComponent<Props> = memo(({ list, label, hideLabel }) => {
  return (
    <div data-testid={`spec-list-${label}`}>
      {!hideLabel && (
        <h4 className="text-sm font-bold mb-2" data-testid="spec-list-label">
          {label}
        </h4>
      )}
      <div className="flex flex-row">
        {list.map((item, idx) => (
          <span
            className="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
            data-testid="spec-list-value"
            key={idx}
          >
            {item}
          </span>
        ))}
      </div>
    </div>
  );
});
