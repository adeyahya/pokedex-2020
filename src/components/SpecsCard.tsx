import { FunctionComponent, memo } from 'react';

interface Props {
  specs: Record<string, React.ReactNode>;
}
export const SpecsCard: FunctionComponent<Props> = memo(
  ({ specs }) => {
    const keys = Object.keys(specs);
    return (
      <div className="flex justify-start flex-wrap" data-testid="specs">
        {keys.map((key) => (
          <div className="flex flex-initial flex-col mr-2 mt-2" key={key}>
            <span className="text-xs font-bold" data-testid="spec-name">
              {key}
            </span>
            <div>
              <span
                className="bg-green-400 p-1 font-bold rounded-lg text-xs"
                data-testid="spec-value"
              >
                {specs[key]}
              </span>
            </div>
          </div>
        ))}
      </div>
    );
  },
  () => true
);
