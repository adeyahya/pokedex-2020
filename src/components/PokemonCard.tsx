import { FunctionComponent, useCallback } from 'react';
import Link from 'next/link';
import { PokemonFragment } from '../queries/types/PokemonFragment';
import { SpecsCard } from './SpecsCard';
import { SpecListCard } from './SpecListCard';
import { useRouter } from 'next/dist/client/router';

interface Props {
  pokemon: PokemonFragment | null;
  prefixUrl?: string;
}

export const PokemonCard: FunctionComponent<Props> = ({ pokemon, prefixUrl = '/' }) => {
  const router = useRouter();
  const targetUrl = prefixUrl.concat(pokemon?.id || '');

  const handleClick = useCallback(() => {
    router.push(targetUrl);
  }, [targetUrl]);

  return (
    <div
      onClick={handleClick}
      className="shadow hover:shadow-lg bg-white rounded-lg cursor-pointer"
      data-testid={`pokemon-card-${pokemon?.name}`}
    >
      <figure className="h-20 md:h-32 lg:h-40 flex items-center justify-center pt-4">
        <img
          className="h-full"
          src={pokemon?.image || ''}
          alt={pokemon?.name || ''}
          data-testid="image"
        />
      </figure>
      <div className="p-4">
        <Link href={targetUrl}>
          <a>
            <h3 className="text-lg font-bold mb-0" data-testid="name">
              {pokemon?.name}
            </h3>
          </a>
        </Link>
        <span className="text-sm italic mb-2 block" data-testid="classification">
          {pokemon?.classification}
        </span>
        <div className="mb-3">
          <SpecListCard list={pokemon?.types || []} label="types" />
        </div>
        <div className="mt-2">
          <SpecsCard
            specs={{
              minWg: pokemon?.weight?.minimum,
              maxWg: pokemon?.weight?.maximum,
              maxHP: pokemon?.maxHP,
              maxCP: pokemon?.maxCP,
            }}
          />
        </div>
      </div>
    </div>
  );
};
