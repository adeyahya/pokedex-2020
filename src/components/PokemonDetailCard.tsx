/* eslint-disable @typescript-eslint/ban-ts-comment */
import Link from 'next/link';
import { FunctionComponent, memo } from 'react';
import { SpecListCard } from './SpecListCard';
import { SpecsCard } from './SpecsCard';
import { PokemonFragment } from '../queries/types/PokemonFragment';

interface Props {
  pokemon: PokemonFragment | null;
}

export const PokemonDetailCard: FunctionComponent<Props> = memo(({ pokemon }) => {
  const fastAttacks: Record<string, string> = (pokemon?.attacks?.fast || []).reduce((acc, curr) => {
    const { name = 'unknown', damage = 'unknown' } = curr || {};
    // @ts-ignore: it still syas that name can be null even after adding default value
    acc[name] = damage?.toString().concat(' Damage');
    return acc;
  }, {} as Record<string, string>);

  const specialAttacks: Record<string, string> = (pokemon?.attacks?.special || []).reduce(
    (acc, curr) => {
      const { name = 'unknown', damage = 'unknown' } = curr || {};
      // @ts-ignore: it still syas that name can be null even after adding default value
      acc[name] = damage?.toString().concat(' Damage');
      return acc;
    },
    {} as Record<string, string>
  );

  return (
    <div data-testid="pokemon-detail-card" className="bg-white rounded-lg overflow-hidden">
      <div className="px-5 grid md:grid-cols-2 gap-4 pt-5 pb-10">
        <div className="flex flex-col justify-center items-center pt-5">
          <figure className="flex flex-center justify-center">
            <img data-testid="image" src={pokemon?.image || ''} alt={pokemon?.name || ''} />
          </figure>
          <div
            data-testid="evolutions"
            className="flex flex-1 w-full flex-row items-center justify-center; md:items-start md:px-5 mt-10"
          >
            {pokemon?.evolutions?.map((evolution, idx) => (
              <Link key={evolution?.name || idx} href={`/pokemon?pid=${evolution?.name || ''}`}>
                <a className="justify-between w-1/2 px-5 md:px-0 md:justify-start md:px-auto md:mr-5">
                  <img
                    data-testid="evolution-image"
                    className="md:h-32"
                    src={evolution?.image || ''}
                    alt={evolution?.name || ''}
                  />
                  <span data-testid="evolution-name" className="text-lg font-bold block mt-2">
                    {evolution?.name}
                  </span>
                </a>
              </Link>
            ))}
          </div>
        </div>
        <section className="md:pl-5">
          <header className="mb-4">
            <h2 data-testid="name" className="text-4xl text-bold">
              {pokemon?.name} <span className="text-gray-500">#{pokemon?.number}</span>
            </h2>
            <span data-testid="classification" className="text-lg italic">
              {pokemon?.classification}
            </span>
          </header>
          <SpecListCard label="Types" list={pokemon?.types || []} />
          <div className="mb-2" />
          <SpecListCard label="Weaknesses" list={pokemon?.weaknesses || []} />
          <div className="mb-2" />
          <SpecListCard label="Resistant" list={pokemon?.resistant || []} />
          <div className="mt-4 mb-4">
            <h4 className="text-xl font-bold">Combat</h4>
            <div className="mb-2" />
            <span className="font-lg font-bold">Fast Attack</span>
            <SpecsCard specs={fastAttacks} />
            <span className="block font-lg font-bold mt-4">Special Attack</span>
            <SpecsCard specs={specialAttacks} />
          </div>
          <div>
            <h4 className="text-xl font-bold">Attributes</h4>
            <SpecsCard
              specs={{
                MaxHp: pokemon?.maxHP,
                MaxCP: pokemon?.maxCP,
                MaxWeight: pokemon?.weight?.maximum,
                MinWeight: pokemon?.weight?.minimum,
                MaxHeight: pokemon?.height?.maximum,
                MinHeight: pokemon?.height?.minimum,
                FleeRate: pokemon?.fleeRate,
              }}
            />
          </div>
        </section>
      </div>
    </div>
  );
});
