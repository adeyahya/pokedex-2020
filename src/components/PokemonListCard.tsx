import { FunctionComponent, useRef, useEffect, memo } from 'react';
import { PokemonCard } from './PokemonCard';
import { PokemonFragment } from '../queries/types/PokemonFragment';

interface Props {
  pokemons: (PokemonFragment | null)[];
  hasMore?: boolean;
  loader?: React.ReactNode;
  onLoadMore?: () => void;
}

export const PokemonListCard: FunctionComponent<Props> = memo(
  ({ pokemons, hasMore, loader, onLoadMore }) => {
    const bottomMarkerNode = useRef<HTMLDivElement | null>(null);
    const observer = useRef<IntersectionObserver | null>(null);

    useEffect(() => {
      if (bottomMarkerNode.current && hasMore && onLoadMore) {
        observer.current = new IntersectionObserver((entries) => {
          const firstEntry = entries[0];
          if (firstEntry.isIntersecting) {
            observer.current?.unobserve(bottomMarkerNode.current as Element);
            onLoadMore();
          }
        });
        observer.current.observe(bottomMarkerNode.current);
      }
    }, [hasMore, onLoadMore]);

    return (
      <div>
        <div className="grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-4 px-5 my-10">
          {pokemons?.map((pokemon, idx) => (
            <PokemonCard key={pokemon?.name || idx} pokemon={pokemon} />
          ))}
          <div ref={bottomMarkerNode} />
        </div>
        {hasMore && loader}
      </div>
    );
  }
);
