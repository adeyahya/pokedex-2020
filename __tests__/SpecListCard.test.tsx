import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import { SpecListCard } from '../src/components/SpecListCard';

const TEST_DATA = [
  [['excepteur', 'ea', 'tempor', 'do', 'ex', 'duis', 'eiusmod']],
  [['cillum', 'eiusmod', 'irure', 'culpa', 'dolore', 'tempor', 'in']],
  [['anim', 'est', 'sit', 'ad', 'veniam', 'duis', 'elit']],
  [['non', 'aliquip', 'dolore', 'deserunt', 'incididunt', 'eiusmod', 'reprehenderit']],
  [['culpa', 'sint', 'dolore', 'exercitation', 'irure', 'veniam']],
  [['occaecat', 'voluptate', 'non', 'nostrud', 'proident']],
];

describe('SpecList Card', () => {
  test.each(TEST_DATA)('renders correctly for case %#', (list) => {
    const label = list[0].concat('-').concat(list[1]);
    render(<SpecListCard list={list} label={label} />);
    expect(screen.getByTestId(`spec-list-${label}`)).toMatchSnapshot();
    expect(
      screen.getByTestId(`spec-list-${label}`).querySelector('[data-testid="spec-list-label"]')
        ?.innerHTML
    ).toEqual(label);
    list.forEach((item) => {
      expect(screen.getByText(item).getAttribute('data-testid')).toEqual('spec-list-value');
    });
  });
});
