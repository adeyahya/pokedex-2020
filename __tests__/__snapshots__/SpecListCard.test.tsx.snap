// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpecList Card renders correctly for case 0 1`] = `
<div
  data-testid="spec-list-excepteur-ea"
>
  <h4
    class="text-sm font-bold mb-2"
    data-testid="spec-list-label"
  >
    excepteur-ea
  </h4>
  <div
    class="flex flex-row"
  >
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      excepteur
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      ea
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      tempor
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      do
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      ex
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      duis
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      eiusmod
    </span>
  </div>
</div>
`;

exports[`SpecList Card renders correctly for case 1 1`] = `
<div
  data-testid="spec-list-cillum-eiusmod"
>
  <h4
    class="text-sm font-bold mb-2"
    data-testid="spec-list-label"
  >
    cillum-eiusmod
  </h4>
  <div
    class="flex flex-row"
  >
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      cillum
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      eiusmod
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      irure
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      culpa
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      dolore
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      tempor
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      in
    </span>
  </div>
</div>
`;

exports[`SpecList Card renders correctly for case 2 1`] = `
<div
  data-testid="spec-list-anim-est"
>
  <h4
    class="text-sm font-bold mb-2"
    data-testid="spec-list-label"
  >
    anim-est
  </h4>
  <div
    class="flex flex-row"
  >
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      anim
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      est
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      sit
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      ad
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      veniam
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      duis
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      elit
    </span>
  </div>
</div>
`;

exports[`SpecList Card renders correctly for case 3 1`] = `
<div
  data-testid="spec-list-non-aliquip"
>
  <h4
    class="text-sm font-bold mb-2"
    data-testid="spec-list-label"
  >
    non-aliquip
  </h4>
  <div
    class="flex flex-row"
  >
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      non
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      aliquip
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      dolore
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      deserunt
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      incididunt
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      eiusmod
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      reprehenderit
    </span>
  </div>
</div>
`;

exports[`SpecList Card renders correctly for case 4 1`] = `
<div
  data-testid="spec-list-culpa-sint"
>
  <h4
    class="text-sm font-bold mb-2"
    data-testid="spec-list-label"
  >
    culpa-sint
  </h4>
  <div
    class="flex flex-row"
  >
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      culpa
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      sint
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      dolore
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      exercitation
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      irure
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      veniam
    </span>
  </div>
</div>
`;

exports[`SpecList Card renders correctly for case 5 1`] = `
<div
  data-testid="spec-list-occaecat-voluptate"
>
  <h4
    class="text-sm font-bold mb-2"
    data-testid="spec-list-label"
  >
    occaecat-voluptate
  </h4>
  <div
    class="flex flex-row"
  >
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      occaecat
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      voluptate
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      non
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      nostrud
    </span>
    <span
      class="mr-2 bg-teal-200 px-2 text-sm font-medium rounded-full"
      data-testid="spec-list-value"
    >
      proident
    </span>
  </div>
</div>
`;
